src/conventional/amdar_bufr2ioda.py

```diff
@@ -0,0 +1,506 @@
+    'northward_wind': ['nlocs'],
+}
+
+metaDataName = iconv.MetaDataName()
+obsValName = iconv.OvalName()
+obsErrName = iconv.OerrName()
+qcName = iconv.OqcName()
+
+AttrData = {
+    'converter': os.path.basename(__file__),
+    'ioda_version': 2,
+    'description': 'Aircraft observations converted from BUFR',
+    'source': 'LDM at NCAR-RAL',
+    'source_files': ''
+}
+
+DimDict = {
+}
+
+float_missing_value = nc.default_fillvals['f4']
+int_missing_value = nc.default_fillvals['i4']
+double_missing_value = nc.default_fillvals['f8']
+long_missing_value = nc.default_fillvals['i8']
+string_missing_value = '_'
+bufr_missing_value =  CODES_MISSING_LONG
+
+iso8601_string = locationKeyList[meta_keys.index('dateTime')][2]
+epoch = datetime.fromisoformat(iso8601_string[14:-1])
+
+
+def main(file_names, output_file):
+
+    # initialize
+    count = [0, 0, 0]
+    start_pos = None
+    start_time = time.time()
+
+    obs_data = {}          # The final outputs.
+    data = {}              # Before assigning the output types into the above.
+    data['eastward_wind'] = []
+    data['northward_wind'] = []
+    data['specific_humidity'] = []
+    data['air_temperature'] = []
+    for key in meta_keys:
+        data[key] = []
+
+    varDict = defaultdict(lambda: DefaultOrderedDict(dict))
+    varAttrs = DefaultOrderedDict(lambda: DefaultOrderedDict(dict))
+
+    for fname in file_names:
+        logging.debug("INFO: Reading file: " + fname)
+        AttrData['source_files'] += ", " + fname
+
+        data, count, start_pos = read_file(fname, count, start_pos, data)
+
+    AttrData['source_files'] = AttrData['source_files'][2:]
+    logging.debug("INFO: All source files: " + AttrData['source_files'])
+
+    if not data:
+        logging.critical("ABORT: no message data was captured, stopping execution.")
+
+    logging.info("--- {:9.4f} BUFR read seconds ---".format(time.time() - start_time))
+
+    nlocs = len(data['dateTime'])
+    DimDict = {'nlocs': nlocs}
+    AttrData['nlocs'] = np.int32(DimDict['nlocs'])
+
+    # Set coordinates and units of the ObsValues.
+    n = 0
+    for iodavar in obsvars:
+        varDict[iodavar]['valKey'] = iodavar, obsValName
+        varDict[iodavar]['errKey'] = iodavar, obsErrName
+        varDict[iodavar]['qcKey'] = iodavar, qcName
+        varAttrs[iodavar, obsValName]['coordinates'] = 'longitude latitude'
+        varAttrs[iodavar, obsErrName]['coordinates'] = 'longitude latitude'
+        varAttrs[iodavar, qcName]['coordinates'] = 'longitude latitude'
+        varAttrs[iodavar, obsValName]['units'] = obsvars_units[n]
+        varAttrs[iodavar, obsErrName]['units'] = obsvars_units[n]
+        varAttrs[iodavar, qcName]['units'] = 'unitless'
+        n += 1
+
+    # Set units of the MetaData variables and all _FillValues.
+    for key in meta_keys:
+        varAttrs[(key, metaDataName)]['units'] = locationKeyList[meta_keys.index(key)][2]
+        if (locationKeyList[meta_keys.index(key)][1] == "string"):
```

Comment:
Suggest using two dictionaries to get the fill value and dtype rather than a bunch of `elif`:
something like
```
missing_vals = {'string' : string_missing_value}
dtypes = {'string': object}
varAttrs[(key, metaDataName)]['_FillValue'] = missing_vals['string']
obs_data[(key, metaDataName)] = np.array(data[key], dtype=dtypes['string'])
``` 
(also what is `object`?)

Reply:
I must be dense, because I am not seeing how your proposed solution provides the right missing and type for ints or floats separately without still using a if-elif block.

Reply:
You can change lines 162-176 to:
```
dtypestr = locationKeyList[meta_keys.index(key)][1]
varAttrs[(key, metaDataName)]['_FillValue'] = missing_vals[dtypestr]
obs_data[(key, metaDataName)] = np.array(data[key], dtype=dtypes[dtypestr])
```

And then at like line 97 do something like:
```
missing_vals = {'float': nc.default_fillvals['f4'], 'string': '_'}
```

Does that make sense?

Reply:
Yup, of course, gotcha and Thanks!